src/utils/RpcFilter.ts

```diff
@@ -0,0 +1,118 @@
+    private readonly validationService: MessageValidationService;
+    private readonly messageCache: MessageCache;
+    private readonly p2pManager: P2PManager;
+
+    constructor(p2pManager: P2PManager) {
+        this.p2pManager = p2pManager;
+        const agreementTime = p2pManager.stateManager.timeConfig.agreementTime;
+        this.validationService = new MessageValidationService(agreementTime);
+        this.messageCache = new MessageCache(agreementTime);
+    }
+
+    /**
+     * Get or create rate limiter for a transport connection
+     */
+    getInboundRateLimiter(address: Address): RateLimiter {
+        let rateLimiter = this.inboundRateLimiters.get(address);
+        if (!rateLimiter) {
+            rateLimiter = createRateLimiter();
+            this.inboundRateLimiters.set(address, rateLimiter);
+        }
+        return rateLimiter;
+    }
+
+    /**
+     * Check RPC message for bandwidth management with signature-based deduplication
+     * @param serializedRpc - The serialized RPC message
+     * @returns Promise<boolean> - true if message should be allowed
+     */
+    async filterRpcMessage(
+        serializedRpc: string,
+        transport: ATransport
+    ): Promise<Rpc | undefined> {
+        try {
+            const dataSizeBytes = Buffer.byteLength(serializedRpc, "utf8");
+            const rpc = deserializeRpc(serializedRpc);
+            const senderAddress = this.p2pManager.profileManager
+                .getProfileByTransport(transport)
+                ?.getEvmAddress();
+
+            if (!senderAddress) {
+                this.p2pManager.disconnectAndBlacklistPeer(transport);
+                return undefined;
+            }
+            if (!rpc) {
+                this.p2pManager.disconnectAndBlacklistPeer(transport);
+                return undefined; // Invalid message format
+            }
+
+            const rpcHash = this.messageCache.isCached(serializedRpc);
+            if (!rpcHash) {
+                const isMalciousTimestamp =
+                    this.validationService.isTimetampFradulent(rpc.timestamp);
+                if (isMalciousTimestamp) {
+                    this.p2pManager.disconnectAndBlacklistPeer(transport);
+                    return undefined;
+                }
+                const isAcceptableTimestamp =
+                    this.validationService.isAcceptableTimestamp(rpc.timestamp);
+                if (!isAcceptableTimestamp) {
+                    return undefined; // Message too old or too far in future
+                }
+                const issuerAddress =
+                    await this.validationService.recoverAddressFromRpc(rpc);
+                if (!issuerAddress) {
+                    this.p2pManager.disconnectAndBlacklistPeer(transport);
+                    return undefined; // Invalid signature
+                }
+                // TODO
+                /*
+                    Solution for general gossip needed here
```

Comment:
So what's missing for general gossip - is a solution here - for now this is not important and the gossip we use is not general (the RPC that uses it, implements the logic)